[Coro] Prebuild a module-level debug info set and share it between all coroutine clones

artempyanykh opened the pull request:
Stacked PRs:
 * #118630
 * #118629
 * __->__#118628


--- --- ---

### [Coro] Prebuild a module-level debug info set and share it between all coroutine clones


Summary:
CoroCloner, by calling into CloneFunctionInto, does a lot of repeated
work priming DIFinder and building a list of common module-level debug
info metadata. For programs compiled with full debug info this can get
very expensive.

This diff builds the data once and shares it between all clones.

Anecdata for a sample cpp source file compiled with full debug info:

|                 | Baseline | IdentityMD set | Prebuilt CommonDI (cur.) |
|-----------------|----------|----------------|--------------------------|
| CoroSplitPass   | 306ms    | 221ms          | 68ms                     |
| CoroCloner      | 101ms    | 72ms           | 0.5ms                    |
| CollectCommonDI | -        | -              | 63ms                     |
| Speed up        | 1x       | 1.4x           | 4.5x                     |

Note that CollectCommonDebugInfo happens once *per coroutine* rather than per clone.

Test Plan:
ninja check-llvm-unit
ninja check-llvm

Compiled a sample internal source file, checked time trace output for scope timings.

TylerNowicki commented:
nit: Perhaps 'Common' is a better word than 'Global'. Global made me think of global variables but I realized in the patch that is not what you are doing. Seems you are just creating debug info based on the original function and sharing that with the continuations / splits?

artempyanykh commented:
> nit: Perhaps 'Common' is a better word than 'Global'. Global made me think of global variables but I realized in the patch that is not what you are doing. Seems you are just creating debug info based on the original function and sharing that with the continuations / splits? Or perhaps just 'CoroDebugInfo'?

Good call @TylerNowicki! I like 'common', updated the stack accordingly. 

It was originally "global" in a sense that it was owned by a module (global) rather than the function (local) and so didn't have to be cloned as part of function cloning. But I agree that it can be confusing this way.

TylerNowicki commented:
LGTM

artempyanykh committed:
[Coro] Prebuild a module-level debug info set and share it between all coroutine clones

Summary:
CoroCloner, by calling into CloneFunctionInto, does a lot of repeated
work priming DIFinder and building a list of common module-level debug
info metadata. For programs compiled with full debug info this can get
very expensive.

This diff builds the data once and shares it between all clones.

Anecdata for a sample cpp source file compiled with full debug info:

|                 | Baseline | IdentityMD set | Prebuilt CommonDI (cur.) |
|-----------------|----------|----------------|--------------------------|
| CoroSplitPass   | 306ms    | 221ms          | 68ms                     |
| CoroCloner      | 101ms    | 72ms           | 0.5ms                    |
| CollectCommonDI | -        | -              | 63ms                     |
| Speed up        | 1x       | 1.4x           | 4.5x                     |

Note that CollectCommonDebugInfo happens once *per coroutine* rather than per clone.

Test Plan:
ninja check-llvm-unit
ninja check-llvm

Compiled a sample internal source file, checked time trace output for scope timings.

stack-info: PR: https://github.com/llvm/llvm-project/pull/118628, branch: users/artempyanykh/fast-coro-upstream/9
